Clean up sw_launch_event.md given current state of launch handling. (#93)
WICG/web-app-launch: sw_launch_event.md

# `launch` Event Explainer

Author: Marijn Kruisselbrink &lt;<mek@chromium.org>&gt;<br>
Former Author: Matt Giuca &lt;<mgiuca@chromium.org>&gt;<br>
Former Author: Eric Willigers &lt;<ericwilligers@chromium.org>&gt;<br>
Former Author: Jay Harris &lt;<harrisjay@chromium.org>&gt;<br>
Former Author: Raymes Khoury &lt;<raymes@chromium.org>&gt;

Created: 2017-09-22

> Updated: 2024-08-28

Reviewer: Wow, been a while!

## Introduction and Motivation

The `launch_handler` manifest field gives web apps limited control over what should happen when a user launches their application. While the currently available options (`navigate-existing`, `navigate-new` and `focus-existing`) address many use cases, there are a few use cases where even more flexibility is desired.

> Both `navigate-existing` and `focus-existing` active the most recently used tab/window of an application, while sometimes an application might want to use a different tab or window to handle the navigation. While this behavior can be

Reviewer: ```suggestion
Both `navigate-existing` and `focus-existing` activate the most recently used tab/window of an application, while sometimes an application might want to use a different tab or window to handle the navigation. While this behavior can be
```

> somewhat emulated with `focus-existing` by having javascript focus or navigate a different existing client, the fact that the "wrong" window gets activated first makes this a less than ideal user experience. In other words, the existing launch handler options handle single-window and arbitrary-many-window apps pretty well, they don't support multiple-but-limited windows.

Reviewer: ```suggestion
partially emulated with `focus-existing` by having javascript focus or navigate an existing client, the fact that the "wrong" window gets activated first makes this a less than ideal user experience. In other words, the existing launch handler options handle single-window and arbitrary-many-window apps pretty well, they don't support multiple-but-limited windows.
```

3. **Protocol Handlers:** A user clicks on a `mailto:` protocol link which a website has registered to handle using the [`registerProtocolHandler`](https://html.spec.whatwg.org/multipage/system-state.html#dom-navigator-registerprotocolhandler) API or [`"protocol_handlers"` manifest member](https://wicg.github.io/manifest-incubations/index.html#protocol_handlers-member).
4. **Web Share Target:** A user shares an image with an Image Editor web app that has registered as
   a share target using the [Web Share Target API](https://wicg.github.io/web-share-target/level-2/).
5. **File Handlers:** A user opens a file that an Image Editor web app has registered to handle using the [`"file_handlers"` manifest member](https://wicg.github.io/manifest-incubations/index.html#file_handlers-member)).

## `launch_handler` manifest member

This explainer proposes adding a new `"service-worker"` value to the `client_mode` field of the `launch_handler` manifest member. This would invoke a `"launch"` event on a service worker instead of handling a launch as normal. This allows the site to for example redirect the navigation into an existing window or trigger a notification.

> Since service worker scopes and web app manifest scopes don't necesarilly match, the service worker used for this event would the active worker of whatever registration has the URL being navigated to/launched in scope. If no such service worker can be found (or if the user agent otherwise decides this particular launch should not trigger a `"launch"` event) the next supported option from the provided list of `client_mode`s will be used instead.

Reviewer: ```suggestion
Since service worker scopes and web app manifest scopes don't necessarily match, the service worker used for this event would be the active worker of whatever registration has the URL being navigated to/launched in scope. If no such service worker can be found (or if the user agent otherwise decides this particular launch should not trigger a `"launch"` event) the next supported option from the provided list of `client_mode`s will be used instead.
```

> Since service worker scopes and web app manifest scopes don't necesarilly match, the service worker used for this event would the active worker of whatever registration has the URL being navigated to/launched in scope. If no such service worker can be found (or if the user agent otherwise decides this particular launch should not trigger a `"launch"` event) the next supported option from the provided list of `client_mode`s will be used instead.

Author: client_mode can already be a fallback chain list according to the spec. The only difference I'm proposing is that currently the spec says that at parse time a user agent should pick the first client_mode it supports, while I think there are cases where we would want to be able to pick the "best" option at run-time since I think we might need more flexibility than "does the user agent support this mode" and might want "does the user agent support this mode in this specific situation".

> Since service worker scopes and web app manifest scopes don't necesarilly match, the service worker used for this event would the active worker of whatever registration has the URL being navigated to/launched in scope. If no such service worker can be found (or if the user agent otherwise decides this particular launch should not trigger a `"launch"` event) the next supported option from the provided list of `client_mode`s will be used instead.

Reviewer 2: oh yeah you're right - it can be a list.

If we don't do this, then many link clicks won't be correctly intercepted.

When a `launch` event is captured as the result of a redirect, the behavior of the underlying page needs to be decided. User agents should have freedom to experiment here, but some options are:
- The tab could be closed. This would be good behavior if a tab was only opened to process the navigation before it could be captured by the app.
- The tab could return to the page it was on prior to the navigation starting. This seems like sensible behavior in the general case, allowing the user to continue on the site they were on prior to clicking the link handled by the launch event. The user may lose state on the site as a result of a navigation happening, but this wouldn't be unexpected since they've clicked a link.

> ### Handling `window.open()`

Reviewer 2: Are we removing this because for all cases that we COULD handle window.open, they all are `noopener` and thus don't return a frame reference?

Author: Not really, I'm mostly removing this because this entire web app launch handling spec merely describes what happens when a launch does happen, not what should or should not be considered a launch. So this seemed off-topic.

Reviewer 2: ACK
